src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -316,4 +694,223 @@ private AIState startBuilding()
                 return AIState.COMPLETE_BUILD;
         }
     }
+
+    private boolean handleMaterials(@NotNull final Block block, @NotNull final IBlockState blockState)
+    {
+        //Breaking blocks doesn't require taking materials from the citizens inventory
+        if (block == Blocks.AIR)
+        {
+            return true;
+        }
+
+        if (isBlockFree(block, block.getMetaFromState(blockState)))
+        {
+            return true;
+        }
+
+        return !checkOrRequestItems(BlockUtils.getItemStackFromBlockState(blockState));
+    }
+
+    /**
+     * Defines blocks that can be built for free.
+     *
+     * @param block    The block to check if it is free.
+     * @param metadata The metadata of the block.
+     * @return true or false.
+     */
+    public static boolean isBlockFree(@Nullable final Block block, final int metadata)
+    {
+        return block == null
+                || BlockUtils.isWater(block.getDefaultState())
+                || block.equals(Blocks.LEAVES)
+                || block.equals(Blocks.LEAVES2)
+                || (block.equals(Blocks.DOUBLE_PLANT) && Utils.testFlag(metadata, 0x08))
+                || (block instanceof BlockDoor && Utils.testFlag(metadata, 0x08))
+                || block.equals(Blocks.GRASS);
+    }
+
+    private void placeBlockAt(@NotNull final Block block, @NotNull final IBlockState blockState, @NotNull final BlockPos coords)
+    {
+        if (block == Blocks.AIR)
+        {
+            worker.setItemStackToSlot(EntityEquipmentSlot.MAINHAND, null);
+
+            if (!world.setBlockToAir(coords))
+            {
+                Log.getLogger().error(String.format("Block break failure at %s", coords));
+            }
+        }
+        else
+        {
+            final Item item = Item.getItemFromBlock(block);
+            worker.setItemStackToSlot(EntityEquipmentSlot.MAINHAND, item == null ? null : new ItemStack(item, 1));
+
+            if (!placeBlock(coords, block, blockState))
+            {
+                Log.getLogger().error(String.format("Block place failure %s at %s", block.getUnlocalizedName(), coords));
+            }
+            worker.swingArm(worker.getActiveHand());
+        }
+    }
+
+    private boolean placeBlock(@NotNull final BlockPos pos, final Block block, @NotNull final IBlockState blockState)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of returns of this method 7, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "placeBlock" is 31 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)